### PR TITLE
Rename datavalues2 to datavalues

### LINKED ISSUE #3795: Tracking new datavalues system
- [x] #3786 
- [x] #3794 
- [x]  #3919 
- [x] Move other functions into new datavalues #4200 
- [x] Move the whole crates which depend on datavalues to datavalues2 and rename datavalues2 to datavalues.  #4200

### LINKED ISSUE #3967: Migrate other functions into datavalues2
Now we have a new designed datavalues crate named `datavalues2` in branch `datavalues-dev`.
This new crate has a better design than the previous `datavalue`, including some new features:

1. New datatypes: we do not use enum-based datatype anymore. Instead, we have a new datatype trait `DataType`, and there is `NullableType` which means the datatype can be nullable.

2. New columns: Corresponding to the new data type and we have `ConstColumn` and `NullableColumn`. The `ConstColumn` is a column with constant value, and the `NullableColumn` is a column with an extra bitmap. The other columns are not nullable by default except `NullableColumn`.

3. `ColumnViewer`:  A helper class that can fast access the column without extra cost. What's more, we do not need to care about whether it is nullable or constant. We can use `value(i)` to access the item by index and `null_at(i)` to check it is null or not. The `ColumnViewer` introduces `null_mask` & `non_const_mask` to better handle nullable and constant cases.

4. `ColumnViewerIter`: A helper class to iterator the column. It's a wrapper of `ColumnViewer`

5. `Scalar` types: includes 10 `primitive types`, `boolean`, and `Vec<u8>` types. It's a wonderful idea from  [type-exercise-in-rust](https://github.com/skyzh/type-exercise-in-rust)

6. `ScalarBinaryExpression` && `ScalarUnaryExpression` based on `Scalar` types.

....

Yet, there are a lot of work TBD to make the new datavalues merge into `main` branch.

The main task is functions, I'll mask the migration as `good-first-issue` cause there are many things to do. It may take long time to finish the migration on one's own. And what's more, it will be helpful to understand the new datavalues under migration.

Tasks are:

- [x] #4029

- [x]  datavalues2 && arithmetic functions  [wip @zhyass]

- [x] #3968

- [x] #3969

- [x] #3970

- [x]  datavalues2 && tuple/cast functions (already done)
- [x]  datavalues2 && hash functions (already done)

- [x] #3971

- [x] #3972 [wip @PsiACE ]

- [x] #3973

- [x] #3974

- [x] #3975

- [x] #3976

- [x] #3977

Some tips on migration:

1. In most of the functions, we do not have to care about the nullable cases, you can take the column as not nullable, see [Function2Adapter](https://github.com/datafuselabs/databend/blob/datavalues-dev/common/functions/src/scalars/function2_adapter.rs)  && [ColumnViewer](https://github.com/datafuselabs/databend/blob/datavalues-dev/common/datavalues2/src/columns/viewer.rs) for more details.

2. In most of the functions, we do not have to care about the constant cases, `ColumnViewer` could handle it well. Also, if all the arguments are constant, it will unwrap the constant column and return a constant column. See [Function2Adapter](https://github.com/datafuselabs/databend/blob/datavalues-dev/common/functions/src/scalars/function2_adapter.rs)  for more details.

3. `ScalarBinaryExpression` && `ScalarUnaryExpression` may be two helpful class to cover most function cases.

4. Good examples:  [hashe functions](https://github.com/datafuselabs/databend/blob/datavalues-dev/common/functions/src/scalars/hashes/hash_base.rs)

5. There will be no more implicit casts, we have to explicitly cast the arguments. Errors during the function's eval will be thrown rather than keeping it as null, this is not compatible with the old datavalues.

6. You can take any un-assigned tasks you like, check out from branch `datavalues-dev` to start. don't forget to modify the unit tests and stateless tests. Good Luck!

### LINKED ISSUE #3972: datavalues2 && math functions

### LINKED ISSUE #3975: datavalues2 && string functions  (some of them are already done, you can take the example in `fn register2`)
- [x] bin
- [x] hex
- [x] oct
- [x] repeat
- [x] mid
- [x] substr
- [x] substring
- [x] substring_index
- [x] unhex
- [x] locate
- [x] position
- [x] instr
- [x] insert
- [x] field
- [x] concat
- [x] replace
- [x] strcmp

Please feel free to comments here to take part of them.

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

## Changelog

 
- Improvement
 

## Related Issues

Fixes #3967
Fixes #3975
Fixes #3972
Fixes #3795 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/functions/src/scalars/maths/abs.rs [hunk 1]
```diff
@@ -47,25 +49,36 @@ impl AbsFunction {
...
+                let s = v as $super;
+                if s == $super::MIN {
+                    ctx.set_error(ErrorCode::Overflow(format!(
+                        "Overflow on abs signed number {}",
+                        v
+                    )));
+                    return $target::default();
+                }
+                $super::abs(s) as $target
+            });
```
Comment (PsiACE) [inline common/functions/src/scalars/maths/abs.rs]: :100:

**Review by PsiACE — APPROVED** [review]

**Comment by PsiACE** [comment]
```
02_0025_function_strings_locate:                                        2022-02-18T10:19:11.352080Z  INFO databend_query::servers::mysql::mysql_handler: MySQL connection coming: Ok(127.0.0.1:33044)
[ FAIL ] - result differs with:
--- /home/runner/work/databend/databend/tests/suites/0_stateless/02_function/02_0025_function_strings_locate.result	2022-02-18 10:17:31.675263144 +0000
+++ /home/runner/work/databend/databend/tests/suites/0_stateless/02_function/02_0025_function_strings_locate.stdout	2022-02-18 10:19:11.436087349 +0000
@@ -1,10 +1,10 @@
 ===LOCATE===
-4
+0
 0
 7
 ===POSITION===
-4
+0
 0
 ===INSTR===
-4
+0
 0
```

**Comment by BohuTANG** [comment]
/lgtm

My chrome is crash when I open the diff :\

File: common/functions/tests/it/scalars/maths/abs.rs [hunk 2]
```diff
@@ -12,41 +12,41 @@
...
             name: "abs(1)",
             columns: vec![Series::from_data([1_u32])],
             expect: Series::from_data(vec![1_u32]),
             error: "",
         },
-        ScalarFunction2Test {
+        ScalarFunctionTest {
             name: "abs(-10086)",
             columns: vec![Series::from_data([-10086])],
             expect: Series::from_data(vec![10086_i32]),
```
Comment (PsiACE) [inline common/functions/tests/it/scalars/maths/abs.rs]:
```suggestion
            expect: Series::from_data(vec![10086_u32]),
```

**Review by PsiACE — CHANGES_REQUESTED** [review]

**Comment by PsiACE** [comment]
thread 'plan_expression_monotonicity::test_dates_function' panicked, emmmm, I dont know why

<details>

```
failures:

---- plan_expression_monotonicity::test_dates_function stdout ----
thread 'plan_expression_monotonicity::test_dates_function' panicked at 'assertion failed: `(left == right)`
  left: `false`,
 right: `true`: f(x) = toStartOfWeek(x+12) is_monotonic', common/planners/tests/it/plan_expression_monotonicity.rs:82:5
stack backtrace:
   0:     0x5607557096fc - std::backtrace_rs::backtrace::libunwind::trace::h508eb199ca9e489d
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x5607557096fc - std::backtrace_rs::backtrace::trace_unsynchronized::h161df61e809db333
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x5607557096fc - std::sys_common::backtrace::_print_fmt::h2770cfc3b1d87576
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x5607557096fc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h563ba30d61935f58
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x560755769b4c - core::fmt::write::h38da4b8ad8c9b0ca
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/fmt/mod.rs:1190:17
   5:     0x5607556f8cf3 - std::io::Write::write_fmt::haf39a6325ce0294c
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/io/mod.rs:1653:15
   6:     0x56075570d9e1 - std::sys_common::backtrace::_print::hf0c3fdfcf4fe3a40
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x56075570d9e1 - std::sys_common::backtrace::print::h7fb9681b360d5f0f
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x56075570d9e1 - std::panicking::default_hook::{{closure}}::hced02a38ced47401
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panicking.rs:295:22
   9:     0x56075570d70b - std::panicking::default_hook::h03f76e2c951cd926
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panicking.rs:311:9
  10:     0x56075570e2a4 - std::panicking::rust_panic_with_hook::hf69c531f680b680b
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panicking.rs:698:17
  11:     0x56075570df87 - std::panicking::begin_panic_handler::{{closure}}::h3509bb68e184c6c5
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panicking.rs:588:13
  12:     0x560755709ba4 - std::sys_common::backtrace::__rust_end_short_backtrace::h9de973f77cb459c7
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/sys_common/backtrace.rs:138:18
  13:     0x56075570dc89 - rust_begin_unwind
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panicking.rs:584:5
  14:     0x5607498a4c43 - core::panicking::panic_fmt::h8d2d16457c4ed0e2
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/panicking.rs:135:14
  15:     0x560755766b6e - core::panicking::assert_failed_inner::h6a087b6b1a487715
  16:     0x56074ec3ab5c - core::panicking::assert_failed::h0568986cd22f0725
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/panicking.rs:174:5
  17:     0x5607498addc8 - it::plan_expression_monotonicity::verify_test::h1d4be1d7ee4a81e1
                               at /home/runner/work/databend/databend/common/planners/tests/it/plan_expression_monotonicity.rs:82:5
  18:     0x5607498cac3e - it::plan_expression_monotonicity::test_dates_function::h476b6d419dad0826
                               at /home/runner/work/databend/databend/common/planners/tests/it/plan_expression_monotonicity.rs:538:9
  19:     0x5607498c93f3 - it::plan_expression_monotonicity::test_dates_function::{{closure}}::h3215a929729aa204
                               at /home/runner/work/databend/databend/common/planners/tests/it/plan_expression_monotonicity.rs:456:1
  20:     0x5607498a5bce - core::ops::function::FnOnce::call_once::h3c4b61a8556e3cb1
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/ops/function.rs:227:5
  21:     0x5607499322e3 - core::ops::function::FnOnce::call_once::h3a93dcaa94f1531e
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/ops/function.rs:227:5
  22:     0x5607499322e3 - test::__rust_begin_short_backtrace::h096dd520a45780b8
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/test/src/lib.rs:575:5
  23:     0x560749930f99 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h06489cf96f38900f
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/alloc/src/boxed.rs:1854:9
  24:     0x560749930f99 - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hc77c5ca894e8562e
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/panic/unwind_safe.rs:271:9
  25:     0x560749930f99 - std::panicking::try::do_call::h3594f77e439e640d
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panicking.rs:492:40
  26:     0x560749930f99 - std::panicking::try::h091388cd8faeb8ce
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panicking.rs:456:19
  27:     0x560749930f99 - std::panic::catch_unwind::h429b19b7f4c5fb2a
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panic.rs:137:14
  28:     0x560749930f99 - test::run_test_in_process::ha335b60864a57f14
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/test/src/lib.rs:598:18
  29:     0x560749930f99 - test::run_test::run_test_inner::{{closure}}::hd4b78ab2d90bb430
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/test/src/lib.rs:492:39
  30:     0x56074993c181 - test::run_test::run_test_inner::{{closure}}::hc5ef8ecc617d9d09
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/test/src/lib.rs:519:37
  31:     0x56074993c181 - std::sys_common::backtrace::__rust_begin_short_backtrace::h6c705132e9c389b0
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/sys_common/backtrace.rs:122:18
  32:     0x5607499042af - std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}::hdec3e20896f52bff
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/thread/mod.rs:498:17
  33:     0x5607499042af - <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h33deef33de17f001
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/panic/unwind_safe.rs:271:9
  34:     0x5607499042af - std::panicking::try::do_call::h58e46747ca22a19e
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panicking.rs:492:40
  35:     0x5607499042af - std::panicking::try::heb3719a0b3929151
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panicking.rs:456:19
  36:     0x5607499042af - std::panic::catch_unwind::h8f3d2517db12ef3d
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/panic.rs:137:14
  37:     0x5607499042af - std::thread::Builder::spawn_unchecked_::{{closure}}::h661f0e22576b8c79
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/thread/mod.rs:497:30
  38:     0x5607499042af - core::ops::function::FnOnce::call_once{{vtable.shim}}::h4085074f99ab9783
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/core/src/ops/function.rs:227:5
  39:     0x56075571a1d3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5040ebd4d05cfbe0
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/alloc/src/boxed.rs:1854:9
  40:     0x56075571a1d3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h9f2a7aba330dbbd9
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/alloc/src/boxed.rs:1854:9
  41:     0x56075571a1d3 - std::sys::unix::thread::Thread::new::thread_start::hdb72b219a1fa43cf
                               at /rustc/f624427f8771c00819684c783bb841bf72095704/library/std/src/sys/unix/thread.rs:108:17
  42:     0x7f50b743c609 - start_thread
  43:     0x7f50b6e7d293 - clone
```

</details>

**Comment by Xuanwo** [comment]
image

Great!

**Comment by sundy-li** [comment]
Just a flaky test #4174  

Let's merge

**Pull request merged**